doc/developer/design/20210811_catalog_only_mode.md

```diff
@@ -0,0 +1,115 @@
+# Catalog-only mode
+
+## Summary
+
+When users have views causing OOM crashes, we want to provide a mechanism that
+still lets users boot into Materialize and take actions to remediate the issue,
+e.g. drop the OOMing view. To accomplish this, we'll introduce a new boot mode
+called "catalog-only mode" that starts Materialize as normal with the exception
+of not shipping any dataflows for user-created objects, such as materialized
+views.
+
+Ideally, we'll also provide a mechanism that lets users enable dataflows for
+their objects while Materialize is running using a statement such as `ALTER
+INDEX <index name> SET ENABLED`.
+
+## Goals
+
+- Support booting in catalog-only mode, which will not:
+    - Ship the dataflow for any user-created object
+    - Support querying user-created objects, such as views
+    - Support inserting into tables
+
+    Any other action should be allowed, such as creating, modifying, or dropping
+    objects.
+- Introduce a `--catalog-only` flag to Materialize that lets users easily turn
+  on catalog-only mode.
+- Add support for `ALTER INDEX ... SET ENABLED` to enable dataflows for objects
```

Comment:
What happens if this index depends on other indexes?

Reply:
Good call! I believe when the optimizer exposes the plan and we check for the validity of the indexes in the plan, we would encounter the enabled or disabled indexes there. Seems like it'll be important to add details to the error message to indicate which indexes could have been enabled to let the query succeed.